include/slate/func.hh

```diff
@@ -209,6 +232,122 @@ transpose_grid(std::function<int(ij_tuple)> old_func)
+///
+/// @param[in] mt
+///     The number of tile rows to consider
+///
+/// @param[in] nt
+///     The number of tile columns to consider
+///
+/// @param[in] func
+///     The tile distribution to inspect
+///
+/// @param[out] order
+///     The GridOrder detected.
+///
+/// @param[out] p
+///     The number of rows in the grid, or -1 of the GridOrder is unknown.
+///
+/// @param[out] q
+///     The number of columns in the grid, or -1 of the GridOrder is unknown.
+///
+/// @return Whether the map is a 2d cyclic grid
+///
+/// @ingroup func
+///
+inline bool is_2d_cyclic_grid(int64_t mt, int64_t nt, std::function<int(ij_tuple)> func,
+                              GridOrder& order, int64_t& p, int64_t& q)
+{
+    if (mt == 0 || nt == 0 || (mt == 1 && nt == 1)) {
+        order = GridOrder::Col;
+        p = 1;
+        q = 1;
+        return true;
+    }
+
+    order = GridOrder::Unknown;
+    p = -1;
+    q = -1;
+
+    GridOrder pred_order = GridOrder::Unknown;
+    if (mt == 1 || nt == 1) {
+        // 1d distribution
+        // Ambiguous layout, so just choose col major
+        pred_order = GridOrder::Col;
+    }
+    else if (func( {1, 0} ) == 1) {
+        pred_order = GridOrder::Col;
+    }
+    else if (func( {0, 1} ) == 1) {
+        pred_order = GridOrder::Row;
+    }
+    else if (func( {1, 0} ) == 0 && func( {0, 1} ) == 0) {
+        pred_order = GridOrder::Col;
+    }
+    else {
+        return false;
+    }
+
+    int64_t pred_p = 0;
+    int64_t pred_q = 0;
+    if (pred_order == GridOrder::Col) {
+        while (pred_p < mt && func( {pred_p, 0} ) == pred_p) {
+            ++pred_p;
+        }
+        while (pred_q < nt && func( {0, pred_q} ) == pred_q*pred_p) {
+            ++pred_q;
+        }
+    }
+    else { // pred_order == GridOrder::Row
+        while (pred_q < nt && func( {0, pred_q} ) == pred_q) {
+            ++pred_q;
+        }
+        while (pred_p < mt && func( {pred_p, 0} ) == pred_p*pred_q) {
+            ++pred_p;
+        }
+    }
+
+    auto ref = process_2d_grid( Layout(pred_order), pred_p, pred_q );
+    for (int64_t i = 0; i < mt; ++i) {
+        for (int64_t j = 0; j < nt; ++j) {
+            if (func( {i, j} ) != ref( {i, j} )) {
+                return false;
+            }
+        }
+    }
+    // Grid is 2d cyclic
+    order = pred_order;
+    p = pred_p;
+    q = pred_q;
+    return true;
+}
+
+//------------------------------------------------------------------------------
+/// Checks whether the given tile map is a 2d cyclic grid (i.e., equivalent
+/// to 'process_2d_grid(layout, p, q)' for some 'layout', 'p', and 'q').
+///
+/// @param[in] mt
+///     The number of tile rows to consider
+///
+/// @param[in] nt
+///     The number of tile columns to consider
+///
+/// @param[in] func
+///     The tile map to inspect
+///
+/// @retval Whether the map is a 2d cyclic grid
+///
+/// @ingroup func
+///
+inline bool is_2d_cyclic_grid(int64_t mt, int64_t nt, std::function<int(ij_tuple)> func)
```

Comment:
Is this function actually used outside the unit tester?

Reply:
No.  My original plan was to remote `gridinfo` and this would have simplified a few asserts.  But, I ended up keeping `gridinfo` for performance reasons and didn't think to remove this.